### PR TITLE
Refactor for DRY state getters_validator_test and getters_test

### PR DESCRIPTION
**Pull request by leolara**
**What type of PR is this?**

Refactor

**What does this PR do? Why is it needed?**

There are many code repetition on state unit tests, this is an example of a way of fixing this

**Which issues(s) does this PR fix?**

Partly #10047

**Other notes for review**

If this approach is seen as good, I will do the same with all the tests in `state` package while extending coverage

### CODE REVIEW

File: beacon-chain/state/testtmpl/getters_validator.go
Comment (terencechain) [inline beacon-chain/state/testtmpl/getters_validator.go:1]:
```suggestion
package testing
```
or `test`

File: beacon-chain/state/testtmpl/getters_validator.go
Comment (terencechain) [inline beacon-chain/state/testtmpl/getters_validator.go:11]:
```suggestion
type getState func() (state.BeaconState, error)
```

File: beacon-chain/state/testtmpl/getters_validator.go
Comment (terencechain) [inline beacon-chain/state/testtmpl/getters_validator.go:13]:
```suggestion
func TestBeaconState_ValidatorAtIndexReadOnly_HandlesNilSlice(
```

File: beacon-chain/state/testtmpl/getters_validator.go
Comment (terencechain) [inline beacon-chain/state/testtmpl/getters_validator.go:15]:
```suggestion
	st getState,
```

File: beacon-chain/state/testtmpl/getters_validator.go
Comment (leolara) [reply beacon-chain/state/testtmpl/getters_validator.go:11]: @terencechain `getState` would be global for the package, and it clash with other factories necessary for other tests templates

File: beacon-chain/state/testtmpl/getters_validator.go
Comment (leolara) [reply beacon-chain/state/testtmpl/getters_validator.go:13]: @terencechain there would be no way of disguising a test from the test template by the function name

File: beacon-chain/state/testtmpl/getters_validator.go
Comment (leolara) [reply beacon-chain/state/testtmpl/getters_validator.go:1]: @terencechain my idea is to create a package not general for testing, but for the specific goal of having all the test code that are repeated in many subpackages of state

File: beacon-chain/state/testtmpl/getters_validator.go
Comment (terencechain) [reply beacon-chain/state/testtmpl/getters_validator.go:11]: It's local within the pkg, and it is better given the function signatures. We can make it more explicit if there really is a need (which I don't believe) later

File: beacon-chain/state/testtmpl/getters_validator.go
Comment (terencechain) [reply beacon-chain/state/testtmpl/getters_validator.go:1]: testing here refers to the test template too; it’s more idiomatic and follows the rest of the repo

**Comment by leolara** [comment]
@terencechain 

So my general idea is to show here an example for a specific pattern that will have to be repeated many times.

As there are v1, v2, v3, state-static/v[1..3] we will have to do this with most of the testing code.

My idea is:
 + to have a specific package for these "test patterns" separated from other test helpers
 + to follow a common pattern on how these are written:
 ++ Use the same name of the function but instead of starting with `Test`  start with `TestTmpl`
 ++ For the factories use as base of the type the name of the test template function, this factory type can be private as only the callee `TestTmpl` needs to use it in its definition, but the caller can just pass the factory function.
 ++ Use the same file name but without `_test` at the end
 ++ This is to avoid clashes of names and make more mechanic the transition to the new DRY system.
 
This idea can be modified of course, but I would like to try to keep these principles of having a mechanical relationship between the names of the repeated tests and the test templates, and have it in its own package separated from other generic test helpers.

**Comment by terencechain** [comment]
> @terencechain
> 
> So my general idea is to show here an example for a specific pattern that will have to be repeated many times.
> 
> As there are v1, v2, v3, state-static/v[1..3] we will have to do this with most of the testing code.
> 
> My idea is:
> 
> * to have a specific package for these "test patterns" separated from other test helpers
> * to follow a common pattern on how these are written:
>   ++ Use the same name of the function but instead of starting with `Test`  start with `TestTmpl`
>   ++ For the factories use as base of the type the name of the test template function, this factory type can be private as only the callee `TestTmpl` needs to use it in its definition, but the caller can just pass the factory function.
>   ++ Use the same file name but without `_test` at the end
>   ++ This is to avoid clashes of names and make more mechanic the transition to the new DRY system.
> 
> This idea can be modified of course, but I would like to try to keep these principles of having a mechanical relationship between the names of the repeated tests and the test templates, and have it in its own package separated from other generic test helpers.

Sorry. Naming is hard, but I'd love to be a little more idiomatic towards Go and around what Prysm has done so far.
 - Most of our templates, setup helpers, test helpers are usually grouped within pkg `testing` under the target directory
 - Instead of starting `Test` or `TestTmpl`. Maybe we can do `VerifyStateErrNilValidators`.  It's cleaner. Keep in mind that we will label this "test-only" via bazel so there's no worry about other code that references this. There won't be any crashing concern here

**Review by terencechain — CHANGES_REQUESTED** [review]
Could you also please run gazelle and create the `BUILD.bazel` file?
1.) bazel run //:gazelle 
2.) add `testonly = True`
3.) edit `visibility` to beacon chain subpkg or even beacon state state subpkg?

**Comment by leolara** [comment]
@terencechain so if we are already doing something similar? Could you send me a few links to the code so I see how it is done in other places and I do the same?

**Comment by terencechain** [comment]
> @terencechain so if we are already doing something similar? Could you send me a few links to the code so I see how it is done in other places and I do the same?

Sure. There should be helpers in the below links that do the same
Screen Shot 2022-01-25 at 5 26 13 PM
.

**Comment by leolara** [comment]
@terencechain please, check again

**Comment by leolara** [comment]
@terencechain @prestonvanloon @rkapka @nisdas  please have a look, I also added BUILD.bazel that I forgot previously

**Comment by leolara** [comment]
> Could you also please run gazelle and create the `BUILD.bazel` file? 1.) bazel run //:gazelle 2.) add `testonly = True` 3.) edit `visibility` to beacon chain subpkg or even beacon state state subpkg?

This is done

**Review by terencechain — DISMISSED** [review]

File: beacon-chain/state/testing/getters.go
Comment (nisdas) [inline beacon-chain/state/testing/getters.go:33]: Maybe a less verbose name here ? ex: `jStateRetriever` or `justifiedStateRetriever`

File: beacon-chain/state/testing/getters.go
Comment (nisdas) [inline beacon-chain/state/testing/getters.go:49]: You do not need to separate out for our native state. They all follow the same interface, you can use the above test for the native state. This applies to all the below following tests.

Comment (leolara) [reply beacon-chain/state/testing/getters.go:49]: @nidas The Native one does not have `clear` method, that is why I indroduced this

Comment (nisdas) [reply beacon-chain/state/testing/getters.go:49]: You can just have the native state clear method to do nothing so as to reduce test duplication here.

File: beacon-chain/state/testing/getters.go
Comment (leolara) [reply beacon-chain/state/testing/getters.go:33]: I wouldn't know that j means justified, and the similar methods start with getState as suggested by @terencechain before as it would follow more the rest of the code.

Also, looking at the name of the tests previously here, I see this level of verbosity in the code for function names

File: beacon-chain/state/testing/getters.go
Comment (leolara) [reply beacon-chain/state/testing/getters.go:49]: If I am not understanding something from your comment, let's fix it in another PR, so this does not get in more conflicts

**Comment by leolara** [comment]
@rkapka it seems it is waiting for your review

File: beacon-chain/state/testing/getters.go
Comment (nisdas) [reply beacon-chain/state/testing/getters.go:49]: Maybe it will be clearer with an example:
```go
func TestBeaconState_MatchCurrentJustifiedCheckpt(t *testing.T) {
	testtmpl.VerifyBeaconState_MatchCurrentJustifiedCheckpt(
		t,
		func(cp *ethpb.Checkpoint) (state.BeaconState, error) {
			return InitializeFromProto(&ethpb.BeaconState{CurrentJustifiedCheckpoint: cp})
		},
func(i state.BeaconState) { 
}
	)
}
```
Basically the clear function is no-op .

**Review by nisdas — APPROVED** [review]
Approving it, as you cant abstract out the native state tests without it failing. We can look at this again later

**Comment by rkapka** [comment]
@leolara We need `bazel run //:gazelle -- fix`

**Comment by leolara** [comment]
> @leolara We need `bazel run //:gazelle -- fix`

Ok

**Comment by leolara** [comment]
> @leolara We need `bazel run //:gazelle -- fix`

Done

**Review by terencechain — APPROVED** [review]

**Comment by leolara** [comment]
@rkapka waiting for your approval

**Review by nisdas — APPROVED** [review]

**Pull request merged**
